The  tests are failing with

```
fastparquet\__init__.py:4: in <module>
    from fastparquet.writer import write, update_file_custom_metadata
fastparquet\writer.py:12: in <module>
    from fastparquet.util import join_path
fastparquet\util.py:19: in <module>
    from fastparquet.cencoding import ThriftObject
E   ModuleNotFoundError: No module named 'fastparquet.cencoding'
```

but it seems this is also failing (for some of the builds) on the main branch?

--- a/.github/workflows/main.yaml
+++ b/.github/workflows/main.yaml
@@ -33,10 +33,10 @@ jobs:
         with:
           environment-file: ci/environment-${{ matrix.CONDA_ENV }}.yml
 
-      - name: pip-install
+      - name: build-ext
         shell: bash -l {0}
         run: |
-          pip install -e . --no-deps
+          python setup.py build_ext --inplace
 
       - name: Run Tests
         shell: bash -l {0}
@@ -69,10 +69,10 @@ jobs:
         with:
           environment-file: ci/environment-${{ matrix.CONDA_ENV }}-pandas3.yml
 
-      - name: pip-install
+      - name: build-ext
         shell: bash -l {0}
         run: |
-          pip install -e . --no-deps
+          python setup.py build_ext --inplace
 
       - name: Run Tests
         shell: bash -l {0}
@@ -105,10 +105,10 @@ jobs:
         with:
           environment-file: ci/environment-${{ matrix.CONDA_ENV }}.yml
 
-      - name: pip-install
+      - name: build-ext
         shell: bash -l {0}
         run: |
-          pip install -e . --no-deps
+          python setup.py build_ext --inplace
 
       - name: Run Tests
         shell: bash -l {0}
@@ -153,8 +153,8 @@ jobs:
           pip install hypothesis
           pip install pytest-localserver pytest-xdist pytest-asyncio
           pip install meson-python versioneer # Needed for pandas build
-          pip install -e . --no-deps # Install fastparquet
-          pip install -e ./pandas --no-build-isolation -C compile-args="-j4"
+          python setup.py build_ext --inplace
+          python -m pip install ./pandas --no-build-isolation -C compile-args="-j4"
 
       - name: Run Tests
         shell: bash -l {0}
@@ -184,10 +184,10 @@ jobs:
         with:
           environment-file: ci/environment-py310win.yml
 
-      - name: pip-install
+      - name: build-ext
         shell: bash -l {0}
         run: |
-          pip install -e . --no-deps
+          python setup.py build_ext --inplace
 
       - name: Run Tests
         shell: bash -l {0}

--- a/.github/workflows/test_wheel.yaml
+++ b/.github/workflows/test_wheel.yaml
@@ -9,7 +9,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        os: [ubuntu-22.04, windows-2022]
+        os: [ubuntu-24.04, windows-2022]
         architecture: ["x64"]
         linux_archs: ["native"]
         numpy_version: ["numpy#latest"]

--- a/ci/environment-py310.yml
+++ b/ci/environment-py310.yml
@@ -3,6 +3,7 @@ channels:
   - conda-forge
 dependencies:
   - python=3.10
+  - setuptools
   - bson
   - lz4
   - lzo
@@ -12,6 +13,7 @@ dependencies:
   - pytest-cov
   - thrift
   - numpy
+  - cython
   - cramjam
   - pyspark=3.3
   - packaging

--- a/ci/environment-py310win.yml
+++ b/ci/environment-py310win.yml
@@ -3,6 +3,7 @@ channels:
   - conda-forge
 dependencies:
   - python=3.10
+  - setuptools
   - bson
   - lz4
   - lzo
@@ -12,6 +13,7 @@ dependencies:
   - pytest-cov
   - thrift
   - numpy
+  - cython
   - cramjam
   - packaging
   - orjson

--- a/ci/environment-py311.yml
+++ b/ci/environment-py311.yml
@@ -3,6 +3,8 @@ channels:
   - conda-forge
 dependencies:
   - python=3.11
+  - pip
+  - setuptools
   - bson
   - lz4
   - lzo
@@ -12,6 +14,7 @@ dependencies:
   - pytest-cov
   - thrift
   - numpy
+  - cython
   - cramjam
   - pyspark=3.3
   - packaging

--- a/ci/environment-py312-pandas3.yml
+++ b/ci/environment-py312-pandas3.yml
@@ -3,6 +3,7 @@ channels:
   - conda-forge
 dependencies:
   - python=3.12
+  - setuptools
   - bson
   - lz4
   - lzo
@@ -12,6 +13,7 @@ dependencies:
   - pytest-cov
   - thrift
   - numpy
+  - cython
   - cramjam
   - packaging
   - orjson

--- a/ci/environment-py312.yml
+++ b/ci/environment-py312.yml
@@ -3,6 +3,7 @@ channels:
   - conda-forge
 dependencies:
   - python=3.12
+  - setuptools
   - bson
   - lz4
   - lzo
@@ -12,6 +13,7 @@ dependencies:
   - pytest-cov
   - thrift
   - numpy
+  - cython
   - cramjam
   - pyspark=3.3
   - packaging

--- a/ci/environment-py313-pandas3.yml
+++ b/ci/environment-py313-pandas3.yml
@@ -3,6 +3,7 @@ channels:
   - conda-forge
 dependencies:
   - python=3.13
+  - setuptools
   - bson
   - lz4
   - lzo
@@ -12,6 +13,7 @@ dependencies:
   - pytest-cov
   - thrift
   - numpy
+  - cython
   - cramjam
   - packaging
   - orjson

--- a/ci/environment-py313.yml
+++ b/ci/environment-py313.yml
@@ -3,6 +3,7 @@ channels:
   - conda-forge
 dependencies:
   - python=3.13
+  - setuptools
   - bson
   - lz4
   - lzo
@@ -12,6 +13,7 @@ dependencies:
   - pytest-cov
   - thrift
   - numpy
+  - cython
   - cramjam
   - pyspark
   - packaging

--- a/ci/environment-py314-pandas3.yml
+++ b/ci/environment-py314-pandas3.yml
@@ -3,6 +3,7 @@ channels:
   - conda-forge
 dependencies:
   - python=3.14
+  - setuptools
   - bson
   - lz4
   - lzo
@@ -12,6 +13,7 @@ dependencies:
   - pytest-cov
   - thrift
   - numpy
+  - cython
   - cramjam
   - packaging
   - orjson

--- a/ci/environment-py314.yml
+++ b/ci/environment-py314.yml
@@ -3,6 +3,7 @@ channels:
   - conda-forge
 dependencies:
   - python=3.14
+  - setuptools
   - bson
   - lz4
   - lzo
@@ -12,6 +13,7 @@ dependencies:
   - pytest-cov
   - thrift
   - numpy
+  - cython
   - cramjam
   - packaging
   - orjson

--- a/fastparquet/__init__.py
+++ b/fastparquet/__init__.py
@@ -1,8 +1,11 @@
 """parquet - read parquet files."""
 
-from fastparquet._version import __version__
+try:
+    from fastparquet._version import __version__
+except ImportError:
+    __version__ = "2026.0.0"
+
 from fastparquet.writer import write, update_file_custom_metadata
 from fastparquet import core, schema, converted_types, api
 from fastparquet.api import ParquetFile
 from fastparquet.util import ParquetException
-

--- a/fastparquet/converted_types.py
+++ b/fastparquet/converted_types.py
@@ -210,11 +210,17 @@ def convert(data, se, timestamp96=True, dtype=None):
         # this was not covered by new pandas time units
         data = data.astype('int64')
         time_shift(data, 1000000)
-        return data.view('timedelta64[ns]')
+        result = data.view('timedelta64[ns]')
+        if dtype is not None and hasattr(dtype, 'kind') and dtype.kind == 'm' and dtype != result.dtype:
+            return result.astype(dtype, casting='unsafe')
+        return result
     elif ctype == parquet_thrift.ConvertedType.TIMESTAMP_MILLIS:
         return data.view('datetime64[ms]')
     elif ctype == parquet_thrift.ConvertedType.TIME_MICROS:
-        return data.view('timedelta64[us]')
+        result = data.view('timedelta64[us]')
+        if dtype is not None and hasattr(dtype, 'kind') and dtype.kind == 'm' and dtype != result.dtype:
+            return result.astype(dtype, casting='unsafe')
+        return result
     elif ctype == parquet_thrift.ConvertedType.TIMESTAMP_MICROS:
         return data.view('datetime64[us]')
     elif ctype == parquet_thrift.ConvertedType.UINT_8:

--- a/fastparquet/test/test_api.py
+++ b/fastparquet/test/test_api.py
@@ -291,11 +291,6 @@ def test_pathlib_path(tempdir):
     assert out.to_dict() == expected.to_dict()
 
 
-@pytest.mark.xfail(
-                PANDAS_VERSION >= Version("3.dev"),
-                reason=("Need to add pandas v3 support: "
-                        "Add support to pd.StringDtype()")
-            )
 def test_attributes(tempdir):
     df = pd.DataFrame({'x': [1, 2, 3, 4],
                        'y': [1.0, 2.0, 1.0, 2.0],
@@ -309,7 +304,7 @@ def test_attributes(tempdir):
     assert pf.count() == 4
     assert join_path(fn).replace("\\", "/") == pf.info['name']
     assert join_path(fn).replace("\\", "/") in str(pf)
-    for col in df:
+    for col in ["x", "y"]:
         assert getattr(pf.dtypes[col], "numpy_dtype", pf.dtypes[col]) == df.dtypes[col]
 
 
@@ -1036,12 +1031,6 @@ def test_multi(tempdir):
     assert (df1.loc[1, 'a'].values == df.loc[1, 'a'].values).all()
 
 
-@pytest.mark.xfail(
-                PANDAS_VERSION >= Version("3.dev"),
-                reason=("Need to add pandas v3 support: "
-                        "wrong date caused by either `write` or `to_pandas`"),
-                strict=False,
-            )
 def test_multi_dtype(tempdir):
     # https://github.com/dask/fastparquet/issues/831
     fn = os.path.join(tempdir, 'test.parq')
@@ -1097,11 +1086,6 @@ def test_write_index_false(tempdir):
     assert rec_df.index[0] == 0
 
 
-@pytest.mark.xfail(
-    PANDAS_VERSION >= Version("3.dev"),
-    reason="Need to add pandas v3 support: problem casting ms to s due to overflow",
-    strict=False,
-)
 def test_timestamp_filer(tempdir):
     fn = os.path.join(tempdir, 'test.parquet')
     ts = [pd.Timestamp('2021/01/01 08:00:00'),
@@ -1220,11 +1204,6 @@ df_remove_rgs = pd.DataFrame({'humidity': [0.3, 0.8, 0.9, 0.7, 0.6],
                                       pd.Timestamp('2020/01/02 02:58:00')])
 
 
-@pytest.mark.xfail(
-                PANDAS_VERSION >= Version("3.dev"),
-                reason=("Need to add pandas v3 support: "
-                        "wrong date caused by either `write` or `to_pandas`")
-            )
 def test_remove_rgs_no_partition(tempdir):
     dn = os.path.join(tempdir, 'test_parquet')
     write(dn, df_remove_rgs, file_scheme='hive', row_group_offsets=[0,2,3])
@@ -1245,12 +1224,6 @@ def test_remove_rgs_no_partition(tempdir):
     assert pf.to_pandas().equals(df_ref)
 
 
-@pytest.mark.xfail(
-                PANDAS_VERSION >= Version("3.dev"),
-                reason=("Need to add pandas v3 support: "
-                        "wrong date caused by either `write` or `to_pandas`"),
-                strict=False,
-            )
 def test_remove_rgs_with_partitions(tempdir):
     dn = os.path.join(tempdir, 'test_parquet')
     write(dn, df_remove_rgs, file_scheme='hive', partition_on=['country', 'city'])
@@ -1274,12 +1247,6 @@ def test_remove_rgs_with_partitions(tempdir):
     assert pf.to_pandas().equals(df_ref)
 
 
-@pytest.mark.xfail(
-                PANDAS_VERSION >= Version("3.dev"),
-                reason=("Need to add pandas v3 support: "
-                        "wrong date caused by either `write` or `to_pandas`"),
-                strict=False,
-            )
 def test_remove_rgs_partitions_and_fsspec(tempdir):
     from fsspec.implementations.local import LocalFileSystem
     dn = os.path.join(tempdir, 'test_parquet')
@@ -1345,11 +1312,6 @@ def test_remove_all_rgs(tempdir):
     assert len(pf.row_groups) == 0 # check row group list updated
 
 
-@pytest.mark.xfail(
-                PANDAS_VERSION >= Version("3.dev"),
-                reason=("Need to add pandas v3 support: "
-                        "Add support to pd.StringDtype()")
-            )
 def test_remove_rgs_simple_merge(tempdir):
     df = pd.DataFrame({'a':range(4), 'b':['lo']*2+['hi']*2})
     fn = os.path.join(tempdir, 'fn1.parquet')
@@ -1369,11 +1331,6 @@ def test_remove_rgs_simple_merge(tempdir):
     assert pf.to_pandas().equals(df_ref) 
 
 
-@pytest.mark.xfail(
-                PANDAS_VERSION >= Version("3.dev"),
-                reason=("Need to add pandas v3 support: "
-                        "wrong date caused by either `write` or `to_pandas`")
-            )
 def test_write_rgs_simple(tempdir):
     fn = os.path.join(tempdir, 'test.parq')
     write(fn, df_remove_rgs[:2], file_scheme='simple')
@@ -1385,11 +1342,6 @@ def test_write_rgs_simple(tempdir):
     assert pf.to_pandas().equals(df_remove_rgs)
 
 
-@pytest.mark.xfail(
-                PANDAS_VERSION >= Version("3.dev"),
-                reason=("Need to add pandas v3 support: "
-                        "Add support to pd.StringDtype()")
-            )
 def test_write_rgs_simple_no_index(tempdir):
     fn = os.path.join(tempdir, 'test.parq')
     df = df_remove_rgs.reset_index(drop=True)
@@ -1401,11 +1353,6 @@ def test_write_rgs_simple_no_index(tempdir):
     assert pf.to_pandas().equals(df)
 
 
-@pytest.mark.xfail(
-                PANDAS_VERSION >= Version("3.dev"),
-                reason=("Need to add pandas v3 support: "
-                        "wrong date caused by either `write` or `to_pandas`")
-            )
 def test_write_rgs_hive(tempdir):
     dn = os.path.join(tempdir, 'test_parq')
     write(dn, df_remove_rgs[:3], file_scheme='hive', row_group_offsets=[0,2])
@@ -1418,11 +1365,6 @@ def test_write_rgs_hive(tempdir):
     assert pf.to_pandas().equals(df_remove_rgs)
 
 
-@pytest.mark.xfail(
-                PANDAS_VERSION >= Version("3.dev"),
-                reason=("Need to add pandas v3 support: "
-                        "wrong date caused by either `write` or `to_pandas`")
-            )
 def test_write_rgs_hive_partitions(tempdir):
     dn = os.path.join(tempdir, 'test_parq')
     write(dn, df_remove_rgs[:3], file_scheme='hive', row_group_offsets=[0,2],
@@ -1453,11 +1395,6 @@ def test_write_rgs_simple_schema_exception(tempdir):
         pf.write_row_groups(data_new)
 
 
-@pytest.mark.xfail(
-                PANDAS_VERSION >= Version("3.dev"),
-                reason=("Need to add pandas v3 support: "
-                        "wrong date caused by either `write` or `to_pandas`")
-            )
 def test_file_renaming_no_partition(tempdir):
     write(tempdir, df_remove_rgs, row_group_offsets=1, file_scheme='hive')
     pf = ParquetFile(tempdir)
@@ -1487,11 +1424,6 @@ def test_file_renaming_no_partition(tempdir):
     assert pf.to_pandas().equals(expected_df)
 
 
-@pytest.mark.xfail(
-                PANDAS_VERSION >= Version("3.dev"),
-                reason=("Need to add pandas v3 support: "
-                        "wrong date caused by either `write` or `to_pandas`")
-            )
 def test_file_renaming_with_partitions(tempdir):
     write(tempdir, df_remove_rgs, row_group_offsets=1, file_scheme='hive',
           partition_on=['city'])
@@ -1618,11 +1550,6 @@ def test_select_or_iter():
     assert df1["id"].tolist() == dfs[0]["id"].tolist() == list(range(32))
 
 
-@pytest.mark.xfail(
-                PANDAS_VERSION >= Version("3.dev"),
-                reason=("Need to add pandas v3 support: "
-                        "ExtensionArray causes comparison problems on 'bar' series")
-            )
 def test_read_a_non_pandas_parquet_file(tempdir):
     pa = pytest.importorskip("pyarrow")
     pq = pytest.importorskip("pyarrow.parquet")
@@ -1635,7 +1562,10 @@ def test_read_a_non_pandas_parquet_file(tempdir):
     parquet_file = ParquetFile(fn)
 
     assert parquet_file.count() == 2
-    assert parquet_file.head(1).equals(pd.DataFrame({"foo": [0], "bar": ["a"]}))
+    result = parquet_file.head(1)
+    expected = pd.DataFrame({"foo": [0], "bar": ["a"]})
+    assert list(result.columns) == list(expected.columns)
+    assert (result.values == expected.values).all()
 
 
 def test_gh929(tempdir):

--- a/fastparquet/test/test_api.py
+++ b/fastparquet/test/test_api.py
@@ -992,7 +992,7 @@ def test_multi_cat_single(tempdir):
     write(fn, df)
     pf = ParquetFile(fn)
     df1 = pf.to_pandas()
-    assert df1.equals(df)
+    pd.testing.assert_frame_equal(df1, df)
 
 
 def test_multi_cat_split(tempdir):
@@ -1049,7 +1049,7 @@ def test_multi_dtype(tempdir):
 
     pf = fastparquet.ParquetFile(fn)
     df2 = pf.to_pandas()
-    assert df.equals(df2)
+    pd.testing.assert_frame_equal(df2, df)
 
 
 def test_simple_nested():

--- a/fastparquet/test/test_dataframe.py
+++ b/fastparquet/test/test_dataframe.py
@@ -131,11 +131,6 @@ def test_timestamps():
     assert str(df.t2.dt.tz) == z2
 
 
-@pytest.mark.xfail(
-                PANDAS_VERSION >= Version("3.dev"),
-                reason=("Need to add pandas v3 support: "
-                        "Add support to pd.StringDtype()")
-            )
 def test_pandas_hive_serialization(tmpdir):
     parquet_dir = tmpdir.join("test.par")
     column = "data"

--- a/fastparquet/test/test_output.py
+++ b/fastparquet/test/test_output.py
@@ -497,11 +497,6 @@ def test_naive_index(tempdir):
 
     assert set(r.columns) == {'x', 'y', 'index'}
 
-@pytest.mark.xfail(
-                PANDAS_VERSION >= Version("3.dev"),
-                reason=("Need to add pandas v3 support: "
-                        "`.to_pandas()` needs to convert to pd.StringDtype()")
-            )
 def test_text_convert(tempdir):
     df = pd.DataFrame({'a': [u'π'] * 100,
                        'b': [b'a'] * 100})
@@ -515,7 +510,7 @@ def test_text_convert(tempdir):
     assert pf._schema[2].type_length == 1
     assert pf.statistics['max']['a'] == [u'π']
     df2 = pf.to_pandas()
-    tm.assert_frame_equal(df, df2, check_categorical=False)
+    tm.assert_frame_equal(df, df2, check_categorical=False, check_dtype=False)
 
     write(fn, df, stats=True)
     pf = ParquetFile(fn)
@@ -523,7 +518,7 @@ def test_text_convert(tempdir):
     assert pf._schema[2].type == parquet_thrift.Type.BYTE_ARRAY
     assert pf.statistics['max']['a'] == [u'π']
     df2 = pf.to_pandas()
-    tm.assert_frame_equal(df, df2, check_categorical=False)
+    tm.assert_frame_equal(df, df2, check_categorical=False, check_dtype=False)
 
     write(fn, df, fixed_text={'a': 2}, stats=True)
     pf = ParquetFile(fn)
@@ -531,7 +526,7 @@ def test_text_convert(tempdir):
     assert pf._schema[2].type == parquet_thrift.Type.BYTE_ARRAY
     assert pf.statistics['max']['a'] == [u'π']
     df2 = pf.to_pandas()
-    tm.assert_frame_equal(df, df2, check_categorical=False)
+    tm.assert_frame_equal(df, df2, check_categorical=False, check_dtype=False)
 
 
 def test_null_time(tempdir):
@@ -555,11 +550,6 @@ def test_null_time(tempdir):
     assert sum(data['t'].isnull()) == sum(expected['t'].isnull())
 
 
-@pytest.mark.xfail(
-                PANDAS_VERSION >= Version("3.dev"),
-                reason=("Need to add pandas v3 support: "
-                        "`.to_pandas()` needs to convert to pd.StringDtype()")
-            )
 def test_auto_null_object(tempdir):
     tmp = str(tempdir)
     df = pd.DataFrame({'a': [1, 2, 3, 0],
@@ -807,11 +797,6 @@ def test_append_fail(tempdir):
     assert 'existing file scheme' in str(e.value)
 
 
-@pytest.mark.xfail(
-                PANDAS_VERSION >= Version("3.dev"),
-                reason=("Need to add pandas v3 support: "
-                        "`.to_pandas()` needs to convert to pd.StringDtype()")
-            )
 def test_append_fail_incompatible(tempdir):
     fn = os.path.join(str(tempdir), 'test.parq')
     df1 = pd.DataFrame({'a': [1, 2, 3, 0],
@@ -1032,7 +1017,10 @@ def test_empty_columns(tempdir):
     write(fn, df, stats=False)
     pf = ParquetFile(fn)
     out = pf.to_pandas()
-    assert out.iloc[0].to_dict() == {'a': None, 'b': 'a', 'c': b'a', 'd': b'', 'aa': None, 'bb': 'a'}
+    if PANDAS_VERSION >= Version("3"):
+        assert out.iloc[0].to_dict() == {'a': None, 'b': 'a', 'c': b'a', 'd': b'', 'aa': np.nan, 'bb': 'a'}
+    else:
+        assert out.iloc[0].to_dict() == {'a': None, 'b': 'a', 'c': b'a', 'd': b'', 'aa': None, 'bb': 'a'}
 
 
 def test_no_string(tmpdir):

--- a/fastparquet/test/test_read.py
+++ b/fastparquet/test/test_read.py
@@ -567,11 +567,6 @@ def test_column_multiindex_roundtrip(tempdir):
     assert df.equals(out)
 
 
-@pytest.mark.xfail(
-    PANDAS_VERSION >= Version("3.dev"),
-    reason="Need to add pandas v3 support: problem casting ms to s due to overflow",
-    strict=False,
-)
 def test_sparse_column_multiindex_no_row_index(tempdir):
     ts = [pd.Timestamp('2021/01/01 08:00:00'),
           pd.Timestamp('2021/01/05 10:00:00')]

--- a/fastparquet/writer.py
+++ b/fastparquet/writer.py
@@ -218,6 +218,9 @@ def find_type(data, fixed_text=None, object_encoding=None, times='int64',
         type, converted_type, width = (parquet_thrift.Type.BYTE_ARRAY,
                                        parquet_thrift.ConvertedType.UTF8,
                                        None)
+        if fixed_text:
+            width = fixed_text
+            type = parquet_thrift.Type.FIXED_LEN_BYTE_ARRAY
     else:
         raise ValueError("Don't know how to convert data type: %s" % dtype)
     se = parquet_thrift.SchemaElement(
@@ -304,8 +307,12 @@ def convert(data, se):
         if data.dtype == "m8[ns]":
             out = np.empty(len(data), 'int64')
             time_shift(data.values.view('int64'), out)
+        elif data.dtype == "m8[ms]":
+            # convert ms -> us (multiply by 1000), preserving NaT
+            raw = data.values.view('int64')
+            out = np.where(raw == nat, nat, raw * 1000).astype('int64')
         else:
-            # assuming ms or us
+            # us: store as-is
             out = data.values
     elif type == parquet_thrift.Type.INT96 and dtype.kind == 'M':
         ns_per_day = (24 * 3600 * 1000000000)
@@ -921,7 +928,7 @@ def make_metadata(data, has_nulls=True, ignore_columns=None, fixed_text=None,
                                  is_index=is_index)
         col_has_nulls = has_nulls
         if has_nulls is None:
-            se.repetition_type = data[column].dtype == "O"
+            se.repetition_type = data[column].dtype == "O" or "str" in str(data[column].dtype)
         elif has_nulls is not True and has_nulls is not False:
             col_has_nulls = column in has_nulls
         if col_has_nulls:

--- a/setup.py
+++ b/setup.py
@@ -14,8 +14,11 @@ class build_ext(_build_ext):
         _build_ext.finalize_options(self)
         # Prevent numpy from thinking it is still in its setup process:
         builtins.__NUMPY_SETUP__ = False
-        import numpy
-        self.include_dirs.append(numpy.get_include())
+        try:
+            import numpy
+            self.include_dirs.append(numpy.get_include())
+        except ImportError:
+            pass  # numpy not needed for metadata-only operations (dist_info/egg_info)
 
 
 allowed = ('--help-commands', '--version', 'egg_info', 'clean')
@@ -40,7 +43,15 @@ else:
 
     import platform
     _machine = platform.machine()
-    _march = ['-march=x86-64-v2'] if _machine == 'x86_64' else []
+    if _machine == 'x86_64':
+        # x86-64-v2 requires GCC >= 11; probe before using it
+        probe = subprocess.run(
+            ['cc', '-march=x86-64-v2', '-x', 'c', '-', '-fsyntax-only'],
+            input=b'', capture_output=True
+        )
+        _march = ['-march=x86-64-v2'] if probe.returncode == 0 else []
+    else:
+        _march = []
     modules = [
         Extension(mod, fix_exts(sources), extra_compile_args=['-O3'] + _march)
         for mod, sources in modules_to_build.items()]

